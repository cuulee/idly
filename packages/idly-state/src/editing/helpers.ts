import { LngLat, Tags } from 'idly-common/lib/osm/structures';
import { EntityLike, NodeLike } from './entityLike';

export type NodeMoveType = (loc: LngLat, entity: NodeLike) => NodeLike;

export const nodeMove: NodeMoveType = (loc, entity) => ({
  ...entity,
  loc,
});

export type TagsModifyType = (tags: Tags, entity: EntityLike) => EntityLike;

export const tagsModify: TagsModifyType = (tags, entity) => ({
  ...entity,
  tags,
});
