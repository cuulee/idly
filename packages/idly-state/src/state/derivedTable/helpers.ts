import { Entity, Relation, Way } from 'idly-common/lib/osm/structures';
import { Derived, DerivedTable } from './index';

export function expandIds(
  ids: IterableIterator<string> | Iterable<string>,
  derivedTable: DerivedTable
): Entity[] {
  const res = [];
  for (const id of ids) {
    const e = derivedTable.get(id) as Derived;
    res.push(e.entity);
  }
  return res;
}

export function entityIdParentWays(
  id: string,
  derivedTable: DerivedTable
): Way[] {
  const d = derivedTable.get(id);

  if (!d) {
    throw new Error('Entity Id' + id + ' not found in derived table');
  }

  return expandIds(d.parentWays, derivedTable) as Way[];
}

export function entityIdParentRelations(
  id: string,
  derivedTable: DerivedTable
): Relation[] {
  const d = derivedTable.get(id);

  if (!d) {
    throw new Error('Entity Id' + id + ' not found in derived table');
  }

  return expandIds(d.parentRelations, derivedTable) as Relation[];
}

export function entitiesMakeParentRelationsLookup(
  ids: Iterable<string> | IterableIterator<string>,
  derivedTable: DerivedTable
): Map<string, Relation[]> {
  const result: Array<[string, Relation[]]> = [];

  for (const id of ids) {
    result.push([id, entityIdParentRelations(id, derivedTable)]);
  }
  return new Map(result);
}
