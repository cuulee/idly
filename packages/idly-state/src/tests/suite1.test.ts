import { nodeFactory, wayFactory } from 'idly-common/lib/osm/entityFactory';
import { EntityType, Way } from 'idly-common/lib/osm/structures';
import {
  OsmState,
  osmStateAddModifieds,
  osmStateAddVirgins,
  osmStateCreate,
  osmStateGetEntity,
  osmStateGetNextId,
  osmStateGetVisible,
  osmStateShred,
} from '../osmState';
import { DerivedTable } from '../osmState/derivedTable';
import { parseFixture } from './helpers';

const derivedTableToJs = (d: DerivedTable) =>
  [...d].map(([k, v]) => [
    k,
    {
      entity: JSON.parse(JSON.stringify(v.entity)),
      parentRelations: [...v.parentRelations],
      parentWays: [...v.parentWays],
    },
  ]);

const getLog = (state: OsmState) => state[1];
const getState = (state: OsmState) => state[0];
const mapToKeys = <T>(map: Map<string, T> | ReadonlyMap<string, T>) => [
  ...map.keys(),
];

describe('from xml to final rendering', () => {
  it('starting snapshots', () => {
    const virginEntities = parseFixture('one.xml');
    const osmState = osmStateCreate();
    osmStateAddVirgins(osmState, virginEntities, '1');
    const state = getState(osmState);
    expect(state.getElementTable().size).toBe(121);
    expect(mapToKeys(state.getElementTable())).toMatchSnapshot();
    expect(state.getMetaTable()).toMatchSnapshot();
    expect(state.getQuadkey('1')).toMatchSnapshot();
    expect(getLog(osmState)).toMatchSnapshot();
  });

  it('first, get visible', () => {
    const virginEntities = parseFixture('one.xml');
    const osmState = osmStateCreate();
    osmStateAddVirgins(osmState, virginEntities, '1');

    const visible1 = osmStateGetVisible(osmState, ['1']);
    expect(mapToKeys(visible1)).toMatchSnapshot();
  });

  it('replaces all of the nodes of ways with other ways nodes', () => {
    const virginEntities = () => parseFixture('one.xml');
    const osmState = osmStateCreate();
    osmStateAddVirgins(osmState, virginEntities(), '123');

    const way: any = virginEntities().find(
      r => r.type === EntityType.WAY && r.id === 'w265149283'
    );

    const newNodes = way.nodes
      .map(id => virginEntities().find(r => r.id === id))
      .map(n =>
        nodeFactory({
          ...n,
          id: osmStateGetNextId(osmState, n.id),
        } as any)
      );

    const newWay = wayFactory({
      ...way,
      id: osmStateGetNextId(osmState, way.id),
      nodes: newNodes.map(r => r.id),
    } as Way);

    const newState = osmStateAddModifieds(osmState, [newWay, ...newNodes]);

    expect(osmStateGetEntity(newState, newWay.id)).toBe(newWay);
    expect(osmStateGetEntity(newState, newNodes[0].id)).toBe(newNodes[0]);

    expect(osmStateShred(newState)).toMatchSnapshot();
  });

  describe('new states should not affect other state"s visibles', () => {
    const virginEntities = () => parseFixture('one.xml');
    const osmState = osmStateCreate();
    osmStateAddVirgins(osmState, virginEntities(), '123');

    const modifiedStates = [osmState];
    const initialVisibles = [
      derivedTableToJs(osmStateGetVisible(osmState, ['123'])),
    ];

    for (let i = 0; i < 10; i++) {
      const prevState = modifiedStates[modifiedStates.length - 1];
      const newNodes = [
        nodeFactory({
          id: osmStateGetNextId(prevState, 'n2708095906'),
          tags: { m: 'nodified' },
        }),
        nodeFactory({
          id: osmStateGetNextId(prevState, 'n2708095924'),
          tags: { m: 'nodified' },
        }),
      ];
      const newState = osmStateAddModifieds(
        modifiedStates[modifiedStates.length - 1],
        newNodes
      );

      modifiedStates.push(newState);

      initialVisibles.push(
        derivedTableToJs(osmStateGetVisible(newState, ['123']))
      );
    }
    it("the previous state's visible should not be altered", () => {
      let count = 0;
      for (const state of modifiedStates) {
        expect(derivedTableToJs(osmStateGetVisible(state, ['123']))).toEqual(
          initialVisibles[count]
        );
        count++;
      }
    });

    it('id should not show up if looking up in previous osm state', () => {
      expect(osmStateGetEntity(osmState, 'n2708095924#3')).toBe(undefined);
    });

    it('the state having the id in its log should show up', () => {
      expect(osmStateGetEntity(modifiedStates[4], 'n2708095924#3')).toEqual(
        nodeFactory({
          id: 'n2708095924#3',
          tags: { m: 'nodified' },
        })
      );
    });

    it('internally the State should be modified', () => {
      expect(osmState[0].getElement('n2708095924#3')).not.toBe(undefined);
    });
  });

  describe('branching the history should not afffect states visibiles ', () => {
    const virginEntities = () => parseFixture('one.xml');
    const osmState = osmStateCreate();
    osmStateAddVirgins(osmState, virginEntities(), '123');

    const initialStates = [osmState];
    const initialVisibles = [
      derivedTableToJs(osmStateGetVisible(osmState, ['123'])),
    ];

    for (let i = 0; i < 4; i++) {
      const prevState = initialStates[initialStates.length - 1];
      const newNodes = [
        nodeFactory({
          id: osmStateGetNextId(prevState, 'n2708095906'),
          tags: { m: 'nodified' },
        }),
        nodeFactory({
          id: osmStateGetNextId(prevState, 'n2708095924'),
          tags: { m: 'nodified' },
        }),
      ];
      const newState = osmStateAddModifieds(
        initialStates[initialStates.length - 1],
        newNodes
      );

      initialStates.push(newState);
      initialVisibles.push(
        derivedTableToJs(osmStateGetVisible(newState, ['123']))
      );
    }

    // branching off from a particular state
    const branchedStates = [initialStates[2]];
    const branchedVisibles = [
      derivedTableToJs(osmStateGetVisible(initialStates[2], ['123'])),
    ];

    for (let i = 0; i < 4; i++) {
      const prevState = branchedStates[branchedStates.length - 1];
      const newNodes = [
        nodeFactory({
          id: osmStateGetNextId(prevState, 'n2708095906'),
          tags: { m: 'nodified' },
        }),
        nodeFactory({
          id: osmStateGetNextId(prevState, 'n2708095924'),
          tags: { m: 'nodified' },
        }),
      ];
      const newState = osmStateAddModifieds(
        branchedStates[branchedStates.length - 1],
        newNodes
      );

      branchedStates.push(newState);
      branchedVisibles.push(
        derivedTableToJs(osmStateGetVisible(newState, ['123']))
      );
    }

    it('the intial states visible should not be altered', () => {
      let count = 0;
      for (const state of initialStates) {
        expect(derivedTableToJs(osmStateGetVisible(state, ['123']))).toEqual(
          initialVisibles[count]
        );
        count++;
      }
    });

    it('the branched states visible should not be altered', () => {
      let count = 0;
      for (const state of branchedStates) {
        expect(derivedTableToJs(osmStateGetVisible(state, ['123']))).toEqual(
          branchedVisibles[count]
        );
        count++;
      }
    });

    it('next id generation should be same in all historic states', () => {
      for (const state of branchedStates) {
        expect(osmStateGetNextId(state, 'n2708095906')).toBe(
          osmStateGetNextId(osmState, 'n2708095906')
        );
      }
      for (const state of initialStates) {
        expect(osmStateGetNextId(state, 'n2708095906')).toBe(
          osmStateGetNextId(osmState, 'n2708095906')
        );
      }
    });
  });
});
