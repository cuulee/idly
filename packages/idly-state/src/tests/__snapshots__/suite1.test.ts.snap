// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`from xml to final rendering first, get visible 1`] = `
Array [
  "n419360106",
  "n419360261",
  "n1511944211",
  "n1511944213",
  "n1511944214",
  "n1511944215",
  "n2657347569",
  "n2657347570",
  "n2657347571",
  "n2657347572",
  "n2657347573",
  "n2657347574",
  "n2657347575",
  "n2657347576",
  "n2657347577",
  "n2657347578",
  "n2657347579",
  "n2657347580",
  "n2657347581",
  "n2657347582",
  "n2657347583",
  "n2657347584",
  "n2657347585",
  "n2657347586",
  "n2657347587",
  "n2657347588",
  "n2657347589",
  "n2657347590",
  "n2708095636",
  "n2708095647",
  "n2708095671",
  "n2708095673",
  "n2708095675",
  "n2708095677",
  "n2708095683",
  "n2708095687",
  "n2708095691",
  "n2708095692",
  "n2708095693",
  "n2708095695",
  "n2708095696",
  "n2708095697",
  "n2708095699",
  "n2708095700",
  "n2708095701",
  "n2708095705",
  "n2708095709",
  "n2708095716",
  "n2708095720",
  "n2708095728",
  "n2708095731",
  "n2708095738",
  "n2708095743",
  "n2708095744",
  "n2708095745",
  "n2708095760",
  "n2708095789",
  "n2708095793",
  "n2708095796",
  "n2708095798",
  "n2708095807",
  "n2708095812",
  "n2708095840",
  "n2708095846",
  "n2708095861",
  "n2708095863",
  "n2708095870",
  "n2708095871",
  "n2708095884",
  "n2708095889",
  "n2708095893",
  "n2708095895",
  "n2708095896",
  "n2708095899",
  "n2708095900",
  "n2708095906",
  "n2708095924",
  "n2708095925",
  "n2708095926",
  "n2708095932",
  "n2708095938",
  "n2708095943",
  "n2708095948",
  "n2708095963",
  "n3232790254",
  "n3232790255",
  "n3232790256",
  "n3232790257",
  "n3232790258",
  "n3232790259",
  "n3232790260",
  "n3232790561",
  "n3232790562",
  "n3232790563",
  "n3232790564",
  "n3232790565",
  "n3232790570",
  "n3232790579",
  "n3232790586",
  "n3232790587",
  "n3232790588",
  "n3232790589",
  "n3232790590",
  "n3232790591",
  "w137883073",
  "w260257909",
  "w260257910",
  "w260257911",
  "w260257912",
  "w265149191",
  "w265149192",
  "w265149202",
  "w265149283",
  "w265149286",
  "w317057463",
  "w317057464",
  "w317057467",
  "w317057468",
  "w317057469",
  "w317057470",
  "w317057473",
]
`;

exports[`from xml to final rendering replaces all of the nodes of ways with other ways nodes 1`] = `
Array [
  State {
    "_elementTable": Map {
      "w265149283#0" => Object {
        "attributes": Object {
          "changeset": "52656111",
          "timestamp": "2017-10-05T19:55:30Z",
          "uid": "290680",
          "user": "wheelmap_visitor",
          "version": "3",
          "visible": true,
        },
        "id": "w265149283#0",
        "nodes": Array [
          "n2708095687#0",
          "n2708095700#0",
          "n2708095716#0",
          "n2708095692#0",
          "n2708095693#0",
          "n2708095697#0",
          "n2708095695#0",
          "n2708095696#0",
          "n2708095701#0",
          "n2708095699#0",
          "n2708095743#0",
          "n2708095744#0",
          "n2708095793#0",
          "n2708095728#0",
          "n2708095675#0",
          "n2708095687#0",
        ],
        "tags": Object {
          "addr:city": "New York",
          "addr:street": "5th Avenue",
          "building": "yes",
          "building:levels": "21",
          "height": "76",
          "name": "The Peninsula New York",
          "nycdoitt:bin": "1034801",
          "start_date": "1905",
          "tourism": "hotel",
          "wheelchair": "limited",
          "wikidata": "Q2413390",
          "wikipedia": "en:The Peninsula New York",
        },
        "type": "way",
      },
      "n2708095687#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095687#0",
        "loc": Object {
          "lat": 40.761486,
          "lon": -73.975461,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095700#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095700#0",
        "loc": Object {
          "lat": 40.7616082,
          "lon": -73.975372,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095716#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095716#0",
        "loc": Object {
          "lat": 40.7616429,
          "lon": -73.9754545,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095692#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095692#0",
        "loc": Object {
          "lat": 40.7615207,
          "lon": -73.9755434,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095693#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095693#0",
        "loc": Object {
          "lat": 40.7615338,
          "lon": -73.9755746,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095697#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095697#0",
        "loc": Object {
          "lat": 40.7615753,
          "lon": -73.9756731,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095695#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095695#0",
        "loc": Object {
          "lat": 40.761564,
          "lon": -73.9756813,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095696#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095696#0",
        "loc": Object {
          "lat": 40.7615744,
          "lon": -73.975706,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095701#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095701#0",
        "loc": Object {
          "lat": 40.7616092,
          "lon": -73.9756806,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095699#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095699#0",
        "loc": Object {
          "lat": 40.7615996,
          "lon": -73.9756577,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095743#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:34Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095743#0",
        "loc": Object {
          "lat": 40.7617089,
          "lon": -73.9755781,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095744#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:34Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095744#0",
        "loc": Object {
          "lat": 40.7617214,
          "lon": -73.9756077,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095793#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:35Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095793#0",
        "loc": Object {
          "lat": 40.7618295,
          "lon": -73.975529,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095728#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095728#0",
        "loc": Object {
          "lat": 40.7616619,
          "lon": -73.9751308,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095675#0" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095675#0",
        "loc": Object {
          "lat": 40.7614209,
          "lon": -73.9753062,
        },
        "tags": Object {},
        "type": "node",
      },
      "w265149283" => Object {
        "attributes": Object {
          "changeset": "52656111",
          "timestamp": "2017-10-05T19:55:30Z",
          "uid": "290680",
          "user": "wheelmap_visitor",
          "version": "3",
          "visible": true,
        },
        "id": "w265149283",
        "nodes": Array [
          "n2708095687",
          "n2708095700",
          "n2708095716",
          "n2708095692",
          "n2708095693",
          "n2708095697",
          "n2708095695",
          "n2708095696",
          "n2708095701",
          "n2708095699",
          "n2708095743",
          "n2708095744",
          "n2708095793",
          "n2708095728",
          "n2708095675",
          "n2708095687",
        ],
        "tags": Object {
          "addr:city": "New York",
          "addr:street": "5th Avenue",
          "building": "yes",
          "building:levels": "21",
          "height": "76",
          "name": "The Peninsula New York",
          "nycdoitt:bin": "1034801",
          "start_date": "1905",
          "tourism": "hotel",
          "wheelchair": "limited",
          "wikidata": "Q2413390",
          "wikipedia": "en:The Peninsula New York",
        },
        "type": "way",
      },
      "n2708095687" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095687",
        "loc": Object {
          "lat": 40.761486,
          "lon": -73.975461,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095700" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095700",
        "loc": Object {
          "lat": 40.7616082,
          "lon": -73.975372,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095716" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095716",
        "loc": Object {
          "lat": 40.7616429,
          "lon": -73.9754545,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095692" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095692",
        "loc": Object {
          "lat": 40.7615207,
          "lon": -73.9755434,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095693" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095693",
        "loc": Object {
          "lat": 40.7615338,
          "lon": -73.9755746,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095697" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095697",
        "loc": Object {
          "lat": 40.7615753,
          "lon": -73.9756731,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095695" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095695",
        "loc": Object {
          "lat": 40.761564,
          "lon": -73.9756813,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095696" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095696",
        "loc": Object {
          "lat": 40.7615744,
          "lon": -73.975706,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095701" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095701",
        "loc": Object {
          "lat": 40.7616092,
          "lon": -73.9756806,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095699" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095699",
        "loc": Object {
          "lat": 40.7615996,
          "lon": -73.9756577,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095743" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:34Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095743",
        "loc": Object {
          "lat": 40.7617089,
          "lon": -73.9755781,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095744" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:34Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095744",
        "loc": Object {
          "lat": 40.7617214,
          "lon": -73.9756077,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095793" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:35Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095793",
        "loc": Object {
          "lat": 40.7618295,
          "lon": -73.975529,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095728" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:33Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095728",
        "loc": Object {
          "lat": 40.7616619,
          "lon": -73.9751308,
        },
        "tags": Object {},
        "type": "node",
      },
      "n2708095675" => Object {
        "attributes": Object {
          "changeset": "20999114",
          "timestamp": "2014-03-09T04:22:32Z",
          "uid": "1764427",
          "user": "lxbarth_nycbuildings",
          "version": "1",
          "visible": true,
        },
        "id": "n2708095675",
        "loc": Object {
          "lat": 40.7614209,
          "lon": -73.9753062,
        },
        "tags": Object {},
        "type": "node",
      },
    },
    "_metaTable": Map {},
    "_quadkeysTable": Map {
      "" => Set {
        "w265149283#0",
        "n2708095687#0",
        "n2708095700#0",
        "n2708095716#0",
        "n2708095692#0",
        "n2708095693#0",
        "n2708095697#0",
        "n2708095695#0",
        "n2708095696#0",
        "n2708095701#0",
        "n2708095699#0",
        "n2708095743#0",
        "n2708095744#0",
        "n2708095793#0",
        "n2708095728#0",
        "n2708095675#0",
      },
    },
  },
  Array [
    Set {
      "w265149283#0",
      "n2708095687#0",
      "n2708095700#0",
      "n2708095716#0",
      "n2708095692#0",
      "n2708095693#0",
      "n2708095697#0",
      "n2708095695#0",
      "n2708095696#0",
      "n2708095701#0",
      "n2708095699#0",
      "n2708095743#0",
      "n2708095744#0",
      "n2708095793#0",
      "n2708095728#0",
      "n2708095675#0",
    },
  ],
]
`;

exports[`from xml to final rendering starting snapshots 1`] = `
Array [
  "n419360106",
  "n419360261",
  "n1511944211",
  "n1511944213",
  "n1511944214",
  "n1511944215",
  "n2657347569",
  "n2657347570",
  "n2657347571",
  "n2657347572",
  "n2657347573",
  "n2657347574",
  "n2657347575",
  "n2657347576",
  "n2657347577",
  "n2657347578",
  "n2657347579",
  "n2657347580",
  "n2657347581",
  "n2657347582",
  "n2657347583",
  "n2657347584",
  "n2657347585",
  "n2657347586",
  "n2657347587",
  "n2657347588",
  "n2657347589",
  "n2657347590",
  "n2708095636",
  "n2708095647",
  "n2708095671",
  "n2708095673",
  "n2708095675",
  "n2708095677",
  "n2708095683",
  "n2708095687",
  "n2708095691",
  "n2708095692",
  "n2708095693",
  "n2708095695",
  "n2708095696",
  "n2708095697",
  "n2708095699",
  "n2708095700",
  "n2708095701",
  "n2708095705",
  "n2708095709",
  "n2708095716",
  "n2708095720",
  "n2708095728",
  "n2708095731",
  "n2708095738",
  "n2708095743",
  "n2708095744",
  "n2708095745",
  "n2708095760",
  "n2708095789",
  "n2708095793",
  "n2708095796",
  "n2708095798",
  "n2708095807",
  "n2708095812",
  "n2708095840",
  "n2708095846",
  "n2708095861",
  "n2708095863",
  "n2708095870",
  "n2708095871",
  "n2708095884",
  "n2708095889",
  "n2708095893",
  "n2708095895",
  "n2708095896",
  "n2708095899",
  "n2708095900",
  "n2708095906",
  "n2708095924",
  "n2708095925",
  "n2708095926",
  "n2708095932",
  "n2708095938",
  "n2708095943",
  "n2708095948",
  "n2708095963",
  "n3232790254",
  "n3232790255",
  "n3232790256",
  "n3232790257",
  "n3232790258",
  "n3232790259",
  "n3232790260",
  "n3232790561",
  "n3232790562",
  "n3232790563",
  "n3232790564",
  "n3232790565",
  "n3232790570",
  "n3232790579",
  "n3232790586",
  "n3232790587",
  "n3232790588",
  "n3232790589",
  "n3232790590",
  "n3232790591",
  "w137883073",
  "w260257909",
  "w260257910",
  "w260257911",
  "w260257912",
  "w265149191",
  "w265149192",
  "w265149202",
  "w265149283",
  "w265149286",
  "w317057463",
  "w317057464",
  "w317057467",
  "w317057468",
  "w317057469",
  "w317057470",
  "w317057473",
]
`;

exports[`from xml to final rendering starting snapshots 2`] = `Map {}`;

exports[`from xml to final rendering starting snapshots 3`] = `
Set {
  "n419360106",
  "n419360261",
  "n1511944211",
  "n1511944213",
  "n1511944214",
  "n1511944215",
  "n2657347569",
  "n2657347570",
  "n2657347571",
  "n2657347572",
  "n2657347573",
  "n2657347574",
  "n2657347575",
  "n2657347576",
  "n2657347577",
  "n2657347578",
  "n2657347579",
  "n2657347580",
  "n2657347581",
  "n2657347582",
  "n2657347583",
  "n2657347584",
  "n2657347585",
  "n2657347586",
  "n2657347587",
  "n2657347588",
  "n2657347589",
  "n2657347590",
  "n2708095636",
  "n2708095647",
  "n2708095671",
  "n2708095673",
  "n2708095675",
  "n2708095677",
  "n2708095683",
  "n2708095687",
  "n2708095691",
  "n2708095692",
  "n2708095693",
  "n2708095695",
  "n2708095696",
  "n2708095697",
  "n2708095699",
  "n2708095700",
  "n2708095701",
  "n2708095705",
  "n2708095709",
  "n2708095716",
  "n2708095720",
  "n2708095728",
  "n2708095731",
  "n2708095738",
  "n2708095743",
  "n2708095744",
  "n2708095745",
  "n2708095760",
  "n2708095789",
  "n2708095793",
  "n2708095796",
  "n2708095798",
  "n2708095807",
  "n2708095812",
  "n2708095840",
  "n2708095846",
  "n2708095861",
  "n2708095863",
  "n2708095870",
  "n2708095871",
  "n2708095884",
  "n2708095889",
  "n2708095893",
  "n2708095895",
  "n2708095896",
  "n2708095899",
  "n2708095900",
  "n2708095906",
  "n2708095924",
  "n2708095925",
  "n2708095926",
  "n2708095932",
  "n2708095938",
  "n2708095943",
  "n2708095948",
  "n2708095963",
  "n3232790254",
  "n3232790255",
  "n3232790256",
  "n3232790257",
  "n3232790258",
  "n3232790259",
  "n3232790260",
  "n3232790561",
  "n3232790562",
  "n3232790563",
  "n3232790564",
  "n3232790565",
  "n3232790570",
  "n3232790579",
  "n3232790586",
  "n3232790587",
  "n3232790588",
  "n3232790589",
  "n3232790590",
  "n3232790591",
  "w137883073",
  "w260257909",
  "w260257910",
  "w260257911",
  "w260257912",
  "w265149191",
  "w265149192",
  "w265149202",
  "w265149283",
  "w265149286",
  "w317057463",
  "w317057464",
  "w317057467",
  "w317057468",
  "w317057469",
  "w317057470",
  "w317057473",
}
`;

exports[`from xml to final rendering starting snapshots 4`] = `Array []`;
